### DISCUSSION TITLE
[development/code] How does setOrUnsetHeadingWithLevel / runRegexReplaceAction work?

### DISCUSSION THREAD

**Discussion by aisbergde** [gsantner/markor#1874] [Q&A] [answered]
I am having trouble understanding in the code how `runRegexReplaceAction` in combination with  `setOrUnsetHeadingWithLevel` should work.

    public static List<ActionButtonBase.ReplacePattern> setOrUnsetHeadingWithLevel(int level)

* `ActionButtonBase.ReplacePattern`  
creates 1 pattern // Replace this exact heading level with nothing  
OK
* `ActionButtonBase.ReplacePattern`  
creates 1 pattern // Replace other headings with commonmark-compatible leading space  
I get the idea: the required Prefix is moved into the suffix, to be not removed in the next 6 (or 7) patterns
* `for (final Pattern pp : AsciidocReplacePatternGenerator.PREFIX_PATTERNS)`
creates 6 patterns (Markdown 7), because there are 6 entries in PREFIX_PATTERNS (Markdown: 7)  
this is also clear: these 6 (or 7) patterns are used, to replace the PREFIX_PATTERNS by the new header

How and when is it ensured that the part of the new header that is appended as a suffix in the second pattern is later moved back to the front? Or more clear: where the suffix later is removed?

At some point and somewhere, yes, the part is moved from the suffix back to the prefix. This works correctly for Markdown, but since I modified the PREFIX_ATX_HEADING a bit for AsciiDoc, it doesn't work correctly with it:

Markdown

    public static final Pattern PREFIX_ATX_HEADING = Pattern.compile("^(\\s{0,3})(#{1,6}\\s)");

AciiDoc

    public static final Pattern PREFIX_ATX_HEADING = Pattern.compile("^(={1,6} {1})");

**Comment by aisbergde** [comment]
Can I perhaps already prepare a pull request, even if there are still open questions? So then you can clarify questions based on WIP code?

**Comment by harshad1** [comment, marked as answer]
So regexReplaceAction runs through a list of search and replace pairs and performs the replace for the first pattern it matches.

It is a way of expressing if - else if - else if with regex statements.

i.e.
if condition 1 matches do replace 1 and terminate
if condition 2 matches do replace 2 and terminate
....

regexReplaceAction will run this on _every selected line_ while chunking the changes and maintaining selection properly etc etc.

**Comment by aisbergde** [comment]
I understand this in General, but I don't understand why it doesn't work in some cases, as I want.
But If I can upload code, which is not yet 100% ready, and it will be advised, then I could postpone the solution.

Should I do this as a Pull Request from my Fork into the Markor Repo? Or should I keep it in my fork, until is is ready?

**Reply by gsantner** [reply]
@aisbergde  
Answering to both of your similar questions:

Yes please make a (work in progress) pull request in whatever state it is.  
Most contributions happen exactly this way, like you can see in the latest example at #1839 . Depending on size it usually it stays open for a while till it's mature enough and is viable to merge 😄 . (Or sometimes it happens that a merge request is superseeded by another, gets obsolete, stalls, ..).

This also comes to the benefit, when you have problems (or we see problems) we can directly refer to specific sections in the code changes by review. 

Generally said, don't worry about git history and merge commits, at Markor all PRs are merged squashed. So one commit containing all changes together, directly on top of master.
